apps/prairielearn/src/batched-migrations/20250214035153_course_instance_usages__submissions__backfill.ts

```diff
@@ -0,0 +1,41 @@
+import { z } from 'zod';
+
+import { makeBatchedMigration } from '@prairielearn/migrations';
+import { loadSqlEquiv, queryAsync, queryRow } from '@prairielearn/postgres';
+
+import { updateCourseInstanceUsagesForSubmission } from '../models/course-instance-usages.js';
+
+const sql = loadSqlEquiv(import.meta.url);
+
+const CUTOFF_DATE = '2025-02-15T00:00:00Z';
+
+export default makeBatchedMigration({
+  async getParameters() {
+    // First delete usage data older than a hard-coded date
+    await queryAsync(sql.delete_old_usages, { CUTOFF_DATE });
+
+    // Only backfill from submissions that are older than the cutoff date
+    const { min, max } = await queryRow(
+      sql.select_bounds,
+      { CUTOFF_DATE },
+      z.object({
+        min: z.bigint({ coerce: true }).nullable(),
+        max: z.bigint({ coerce: true }).nullable(),
+      }),
+    );
+    return { min, max, batchSize: 1000 };
+  },
+  async execute(start: bigint, end: bigint): Promise<void> {
+    for (let i = start; i <= end; i++) {
```

Comment:
I get that you want to reuse `updateCourseInstanceUsagesForSubmission` as is, but my gosh, it's going to be slow to run >300 million sequential queries in prod. Even if in the _absolute best case_ it takes 1ms per iteration, we're looking at this batched migration taking ~3.5 days to execute. In practice I expect it to be much longer: just `select_user_id_for_submission_id` on its own takes >1ms to execute, ignoring network round trips and other overhead. I would very very strongly encourage you to write a bespoke implementation of the usage updating query that can itself work in batches instead of doing things serially as they're done here.

Reply:
Changed to a batched query in https://github.com/PrairieLearn/PrairieLearn/pull/11366/commits/c63c04ef9edf429cb978d84880faa45d5daec4f9